ibis/expr/api.py

```diff
@@ -4193,22 +4193,57 @@ def mutate(table, exprs=None, **mutations):
         for name, expr in sorted(mutations.items(), key=operator.itemgetter(0))
     )
 
+    # The below logic computes the mutation node exprs by splitting the
+    # assignment exprs into two disjoint sets:
+    # 1) overwriting_cols_to_expr, which maps a column name to its expr
+    # if the expr contains a column that overwrites an existing table column
+    # 2) non_overwriting_exprs, which is a list of all exprs that do not do
+    # any overwriting.
+    # Given these two data structures, we can compute the mutation node exprs
+    # based on whether any columns are being overwritten.
+    overwriting_cols_to_expr = {}
+    non_overwriting_exprs = []
+    table_schema = table.schema()
     for expr in exprs:
-        if expr.get_name() and isinstance(expr, ir.DestructColumn):
-            raise com.ExpressionError(
-                f"Cannot name a destruct column: {expr.get_name()}"
-            )
+        is_first_overwrite = True
+        expr_contains_overwrite = False
+        if isinstance(expr, ir.DestructColumn):
+            if expr.get_name():
+                raise com.ExpressionError(
+                    f"Cannot name a destruct column: {expr.get_name()}"
+                )
+            for name in expr.type().names:
+                if name in table_schema:
+                    # The below is necessary to ensure that:
+                    # A) all overwritten cols inside the DestructColumn are
+                    # accounted for, while
+                    # B) we don't repeat the same DestructColumn expr more
+                    # than once inside the final mutation node exprs.
+                    # This is both okay and necessary because DestructColumn
```

Comment:
umm maybe dumb question, but why do we want to allow this at all? meaning if we destructure why are we allowing *overwriting* shouldn't this be the clients issue?

Reply:
I'm not sure how a client would be able to work around this issue since ibis creates the DestructColumn and the expression would fail inside ibis mutate at creation-time without logic to handle this case.

Reply:
my point is that if we assign with a destructured column that overwrites existing columns we should simply raise rather than have this work. This seems extra complicated and ordering is very important here. A bit fragile. Why do we care about supporting this case? the user can easily just de-select that column. This could lead to unexpected results, e.g. *some* destructured columns overwrite existing but some do not, that is user error.